addons/binding/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java

```diff
@@ -344,47 +286,131 @@ private SamsungTvService findServiceInstance(String serviceName) {
+    private synchronized void startService(SamsungTvService service) {
         if (service != null) {
             service.addEventListener(this);
             service.start();
             services.add(service);
         }
     }
 
-    private void stopService(SamsungTvService service) {
+    private synchronized void stopService(SamsungTvService service) {
         if (service != null) {
             service.stop();
             service.removeEventListener(this);
             services.remove(service);
         }
     }
 
-    private void stopServices() {
-        logger.debug("Shutdown all Samsung services");
-        for (SamsungTvService service : services) {
-            stopService(service);
+    @Override
+    public void thingDiscovered(DiscoveryService source, DiscoveryResult result) {
+
+        if (configuration.hostName.equals(result.getProperties().get(SamsungTvConfiguration.HOST_NAME))) {
+            logger.debug("thingDiscovered: {}, {}", result.getProperties().get(SamsungTvConfiguration.HOST_NAME),
+                    result);
+            /*
+             * SamsungTV discovery services creates thing UID from UPnP UDN.
+             * When thing is generated manually, thing UID may not match UPnP UDN, so store it for later use (e.g.
+             * thingRemoved).
+             */
+            upnpThingUID = result.getThingUID();
+            logger.debug("thingDiscovered, thingUID={}, discoveredUID={}", this.getThing().getUID(), upnpThingUID);
+            checkAndCreateServices();
         }
-        services.clear();
     }
+
+    @Override
+    public void thingRemoved(DiscoveryService source, ThingUID thingUID) {
+        if (thingUID.equals(upnpThingUID)) {
+            logger.debug("Thing Removed: {}", thingUID);
+            shutdown();
+            putOffline();
+        }
+    }
+
+    @Override
+    public @Nullable Collection<@NonNull ThingUID> removeOlderResults(DiscoveryService source, long timestamp,
+            @Nullable Collection<@NonNull ThingTypeUID> thingTypeUIDs, @Nullable ThingUID bridgeUID) {
+        return null;
+    }
+
+    /**
+     * Send multiple WOL packets spaced with 100ms intervals and resend command
+     *
+     * @param channel Channel to resend command on
+     * @param command Command to resend
+     */
+    private void sendWOLandResendCommand(String channel, Command command) {
+
+        if (configuration.macAddress == null || configuration.macAddress.isEmpty()) {
+            logger.warn("Cannot send WOL packet to {} MAC address unknown", configuration.hostName);
+            return;
+        }
+
+        logger.info("Send WOL packet to {} ({})", configuration.hostName, configuration.macAddress);
+
+        // send max 10 WOL packets with 100ms intervals
+        scheduler.schedule(new Runnable() {
+            int count = 0;
+
+            @Override
+            public void run() {
+                count++;
+                if (count < 10) {
```

Comment:
10 because of UDP? Is it common to do it like this?

Reply:
There are some references/reports that you need to repeat the WOL packet multiple times for it to be effective. This could be due to the nature of UDP or because the receiver is in dep sleep and may miss some packets.

I will change the hard-coded 10 with a constant for the moment or should we make it configurable?

Reply:
constant would be ok for me as well